File: 01-Programming-Languages/Go/Data-Structures/Tree/binary_tree.go
package main

import "fmt"

type TreeNode struct {
    Value int
    Left  *TreeNode
    Right *TreeNode
}

func (t *TreeNode) Insert(value int) *TreeNode {
    if t == nil {
        return &TreeNode{Value: value}
    }
    
    if value < t.Value {
        t.Left = t.Left.Insert(value)
    } else if value > t.Value {
        t.Right = t.Right.Insert(value)
    }
    
    return t
}

func (t *TreeNode) Inorder() {
    if t != nil {
        t.Left.Inorder()
        fmt.Printf("%d ", t.Value)
        t.Right.Inorder()
    }
}

func (t *TreeNode) Preorder() {
    if t != nil {
        fmt.Printf("%d ", t.Value)
        t.Left.Preorder()
        t.Right.Preorder()
    }
}

func (t *TreeNode) Postorder() {
    if t != nil {
        t.Left.Postorder()
        t.Right.Postorder()
        fmt.Printf("%d ", t.Value)
    }
}

func (t *TreeNode) Search(value int) bool {
    if t == nil {
        return false
    }
    
    if value == t.Value {
        return true
    } else if value < t.Value {
        return t.Left.Search(value)
    } else {
        return t.Right.Search(value)
    }
}

func main() {
    var root *TreeNode
    values := []int{50, 30, 70, 20, 40, 60, 80}
    
    for _, v := range values {
        root = root.Insert(v)
    }
    
    fmt.Print("Inorder: ")
    root.Inorder()
    fmt.Println()
    
    fmt.Print("Preorder: ")
    root.Preorder()
    fmt.Println()
    
    fmt.Print("Postorder: ")
    root.Postorder()
    fmt.Println()
    
    fmt.Println("Search 40:", root.Search(40))
    fmt.Println("Search 90:", root.Search(90))
}

File: 01-Programming-Languages/R/Data-Structures/Tree/binary_tree.R
TreeNode <- setRefClass("TreeNode",
  fields = list(
    value = "numeric",
    left = "ANY",
    right = "ANY"
  ),
  methods = list(
    initialize = function(value) {
      .self$value <- value
      .self$left <- NULL
      .self$right <- NULL
    }
  )
)

BinaryTree <- setRefClass("BinaryTree",
  fields = list(
    root = "ANY"
  ),
  methods = list(
    initialize = function() {
      .self$root <- NULL
    },
    
    insert = function(value) {
      new_node <- TreeNode$new(value)
      if (is.null(.self$root)) {
        .self$root <- new_node
        return()
      }
      current <- .self$root
      repeat {
        if (value < current$value) {
          if (is.null(current$left)) {
            current$left <- new_node
            return()
          }
          current <- current$left
        } else if (value > current$value) {
          if (is.null(current$right)) {
            current$right <- new_node
            return()
          }
          current <- current$right
        } else {
          return()  # Duplicate
        }
      }
    },
    
    inorder = function(node = .self$root) {
      if (is.null(node)) return()
      .self$inorder(node$left)
      cat(node$value, " ")
      .self$inorder(node$right)
    },
    
    search = function(value) {
      current <- .self$root
      while (!is.null(current)) {
        if (value == current$value) return(TRUE)
        if (value < current$value) {
          current <- current$left
        } else {
          current <- current$right
        }
      }
      FALSE
    }
  )
)

tree <- BinaryTree$new()
for (v in c(50, 30, 70, 20, 40, 60, 80)) {
  tree$insert(v)
}

cat("Inorder: ")
tree$inorder()
cat("\n")

cat("Search 40:", tree$search(40), "\n")
cat("Search 90:", tree$search(90), "\n")